p2pclient/dht.go

```diff
@@ -0,0 +1,350 @@
+package p2pclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+
+	ggio "github.com/gogo/protobuf/io"
+	cid "github.com/ipfs/go-cid"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	pb "github.com/libp2p/go-libp2p-daemon/pb"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// PeerInfo wraps the PeerInfo message from our protobuf with richer types.
+type PeerInfo struct {
+	// Id is the peer's ID
+	ID peer.ID
+	// Addrs are the peer's listen addresses.
+	Addrs []ma.Multiaddr
+}
+
+func convertPbPeerInfo(pbi *pb.PeerInfo) (*PeerInfo, error) {
+	if pbi == nil {
+		return nil, errors.New("null peerinfo")
+	}
+
+	id, err := peer.IDFromBytes(pbi.GetId())
+	if err != nil {
+		return nil, err
+	}
+
+	addrs := make([]ma.Multiaddr, 0, len(pbi.Addrs))
+	for _, addrbytes := range pbi.Addrs {
+		addr, err := ma.NewMultiaddrBytes(addrbytes)
+		if err != nil {
+			return nil, err
```

Comment:
i think we should just log and eat this conversion error -- there may be multiaddrs from protocols we don't know how to parse.